yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -62,10 +62,23 @@ private[spark] class ApplicationMaster(
     .asInstanceOf[YarnConfiguration]
   private val isClusterMode = args.userClass != null
 
-  // Default to numExecutors * 2, with minimum of 3
-  private val maxNumExecutorFailures = sparkConf.getInt("spark.yarn.max.executor.failures",
-    sparkConf.getInt("spark.yarn.max.worker.failures",
-      math.max(sparkConf.getInt("spark.executor.instances", 0) *  2, 3)))
+  // Default to numExecutors * 2 (maxExecutors in the case that we are
+  // dynamically allocating executors), with minimum of 3.
+  private val maxNumExecutorFailures =
+    sparkConf.getInt("spark.yarn.max.executor.failures",
+      sparkConf.getInt("spark.yarn.max.worker.failures",
+        math.max(
+          3,
+          2 * sparkConf.getInt(
+            if (Utils.isDynamicAllocationEnabled(sparkConf))
+              "spark.dynamicAllocation.maxExecutors"
```

Comment:
If "spark.dynamicAllocation.maxExecutors" is not set, it is still 3 as minimum. So from your change, user have to configure "spark.dynamicAllocation.maxExecutors" to increase max failure attempts if dynamic allocation is enabled. So in that way why not directly configure "spark.yarn.max.worker.failures", that will be more straightforward?